source/slang/slang-ir-header-export.cpp

```diff
@@ -0,0 +1,110 @@
+// slang-ir-header-export.cpp
+#include "slang-ir-header-export.h"
+
+#include "slang-ir.h"
+#include "slang-ir-insts.h"
+
+namespace Slang
+{
+
+static void markForHeaderExport_(IRBuilder & builder, IRInst * inst)
+{
+    if (auto func = as<IRFunc>(inst))
+    {
+        if (!func->findDecoration<IRHeaderExportDecoration>())
+        {    
+            builder.addHeaderExportDecoration(func);
+        }
+        for (auto param : func->getFirstBlock()->getParams())
+        {
+            markForHeaderExport_(builder, param->getDataType());
+        }
+    }
+    else if (auto type = as<IRType>(inst))
+    {
+        // If it's a basic type, we're done.
+        if (as<IRBasicType>(type) || as<IRVoidType>(type))
+            return;
+        
+        // If it's a struct type, mark for py-export.
+        if (auto structType = as<IRStructType>(type))
+        {
+            IRBuilder builder(structType->getModule());
+
+            // If it already has a header-export decoration, we're done.
+            if (!structType->findDecoration<IRHeaderExportDecoration>())
+            {    
+                builder.addHeaderExportDecoration(structType);
+            }
+
+            for (auto field : structType->getFields())
+            {
+                markForHeaderExport_(builder, field->getFieldType());
+            }
+            return;
+        }
+        else if (auto arrayType = as<IRArrayType>(type))
+        {
+            // TODO: not sure about this
+            IRBuilder builder(arrayType->getModule());
+            if (!arrayType->findDecoration<IRHeaderExportDecoration>())
+                builder.addHeaderExportDecoration(arrayType);
+
+            markForHeaderExport_(builder, arrayType->getElementType());
+            return;
+        }
```

Comment:
I like to suggest to write some test cases for this part, then it will become more clear of what needs to happen.